[1.19.x] Make LivingSetAttackTargetEvent compatible with the Brain/Behavior system. (Port of PR #8918)

Budschie opened the pull request:
This PR makes the `LivingSetAttackTargetEvent` compatible with the behavior system. This allows modders to check which entities are being attacked by axolotls, zoglins, hoglins and piglins (which was not possible previously). The example mod included in this PR allows you to be ignored by piglins when holding a stick in your main hand.

The motivation of creating this PR was that I wanted players not to get attacked by piglins whenever they were morphed into one using a morph mod. Achieving this was not possible by just using events, hence I created this PR.

Budschie committed:
Ported PR #8918 to 1.19.2

eerussianguy commented:
A couple QOL ideas, not blockers:

- Have a helper method in the event that sets the target in the correct memory
- Have a helper method that erases the target
- Have a helper method called like getPlayerTarget() that returns null if the target is not a player

Budschie commented:
> A couple QOL ideas, not blockers:
> 
> * Have a helper method in the event that sets the target in the correct memory
> 
> * Have a helper method that erases the target
> 
> * Have a helper method called like getPlayerTarget() that returns null if the target is not a player

Good idea! I will probably also replace the LivingSetAttackTargetEvent\#isCausedByBehavior by something more generic like an enum (of course this enum will implement an interface because of extensibility).

Budschie committed:
Introduced LivingSetAttackTargetEvent#setTarget

Budschie committed:
Added LivingSetAttackTargetEvent#getCurrentTarget

Budschie committed:
Added sanity checks for getTarget and getCurrentTarget

Budschie commented:
> A couple QOL ideas, not blockers:
> 
> * Have a helper method in the event that sets the target in the correct memory
> 
> * Have a helper method that erases the target
> 
> * Have a helper method called like getPlayerTarget() that returns null if the target is not a player

I have now successfully implemented a `LivingSetAttackTargetEvent#setTarget` method which allows you to set a new target or erase the target (by providing `null` as an input to that method).

diesieben07 commented on src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java:
```diff
@@ -27,16 +35,168 @@
+    // and do all of this like it is done in the ComputeFovModifierEvent. By that time, most mod developers will
+    // (hopefully) use LivingSetAttackTargetEvent#setTarget instead of setting the target on their own, which will
+    // minimize the impact of such a breaking change.
+    public static interface ILivingTargetType
```
Why does this need to be public and therefor be subject to breaking changes? I think this interface should just be gone, the enum made private and used directly internally.
To then allow creating the event with the two different types you can introduce static factory methods.

Budschie replied:
The reason that I used an interface is to make the event extensible for other mods (that maybe add new ways of setting a Target). However, after thinking about it a bit, I must admit that this will probably never happen, so, yeah, I should probably remove the interface.

diesieben07 reviewed (changes requested):
Another problem that I have noticed: 
Your call to `setTarget` will cause `LivingSetAttackTarget` to be fired again, offering up a big risk for infinite recursion.
Additionally, due to the way the attack code works, preventing attacks via LivingSetAttackTarget is problematic.
Considering the following scenario:
- A Mob can attack players
- There are 2 players nearby
- Mob AI picks player A to attack
- LivingSetAttackTarget is fired and a mod prevents the attack by using setTarget(null)
- random timeout happens (using `NearestAttackableTargetGoal` as an example)
- Mob AI picks player A to attack again

=> Mob never picks a target to attack, even though it could attack player B

Budschie commented:
> Another problem that I have noticed: Your call to `setTarget` will cause `LivingSetAttackTarget` to be fired again, offering up a big risk for infinite recursion. Additionally, due to the way the attack code works, preventing attacks via LivingSetAttackTarget is problematic. Considering the following scenario:
> 
> * A Mob can attack players
> 
> * There are 2 players nearby
> 
> * Mob AI picks player A to attack
> 
> * LivingSetAttackTarget is fired and a mod prevents the attack by using setTarget(null)
> 
> * random timeout happens (using `NearestAttackableTargetGoal` as an example)
> 
> * Mob AI picks player A to attack again
> 
> 
> => Mob never picks a target to attack, even though it could attack player B

Yep, that's problematic, and I've come to the conclusion that I should probably remove the stuff about setting targets from this PR and move that into a separate PR (and a separate event).

Budschie committed:
Removed LivingSetAttackTargetEvent#setTarget/getCurrentTarget.

Budschie commented:
> => Mob never picks a target to attack, even though it could attack player B

I wonder if this behavior can be avoided. 
It should be pretty doable to integrate something like a `TargetSelectedEvent` into the `NearestAttackableTargetGoal` (although there are a few other target goals in which I would have to hook into, for example the `DefendVillageTargetGoal`), yet I fear that Minecraft's behavior part of the target selection code is designed in a way in which Forge cannot easily hook into.

The `StartAttacking` behavior is simply provided with a Function that can be used to determine the entity which should be attacked; there is no single target selection method to hook into. Instead, I would have to hook into every entities' target selection code separately, which *is* feasible, but it is a high maintenance hassle nonetheless as you would have to check every single MC update for new entities which use the `StartAttacking` behavior. Additionally, it would also put the burden on supporting this event for modded entities on the modders, which is also not great. 

So, all in all, I think that finding an elegant solution to this problem is rather challenging; I would be happy if anyone has any ideas or solutions regarding the difficulties mentioned above.

Budschie committed:
Added missing newline at end of file

Budschie committed:
Adjusted documentation in LivingSetAttackTargetEvent

Budschie committed:
Added LivingChangeTargetEvent

Budschie committed:
Renamed "GOAL_TARGET" to "MOB_TARGET" (LivingTargetType enum)

Budschie committed:
Extended test mod

ChampionAsh5357 commented on src/main/java/net/minecraftforge/event/entity/living/LivingChangeTargetEvent.java:
```diff
@@ -0,0 +1,74 @@
+        return originalTarget;
+    }
+    
+    @Override
+    public boolean isCancelable()
+    {
+        return true;
+    }
```
Use `@Cancelable` annotation, not overriding method on event.

Budschie replied:
Fixed: 214cbd5abc57e331cc7397b1e753bf76a7de625e

ChampionAsh5357 commented on src/main/java/net/minecraftforge/event/entity/living/LivingChangeTargetEvent.java:
```diff
@@ -0,0 +1,74 @@
+ * This event allows you to change the target an entity has. <br>
+ * This event is fired before {@link LivingSetAttackTargetEvent}. <br>
+ * <br>
+ * This event is fired via the {@link ForgeHooks#}
```
Not finished javadoc.

Budschie replied:
Fixed: c8274622fe30303ec1028dfd70d3841605bb9459

ChampionAsh5357 commented on src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java:
```diff
@@ -27,16 +28,48 @@
+    
+    public static enum LivingTargetType implements ILivingTargetType
+    {
+        MOB_TARGET,
+        BEHAVIOR_TARGET;
```
Documentation on the differences between the two targets.

Budschie replied:
Fixed: c8274622fe30303ec1028dfd70d3841605bb9459

ChampionAsh5357 commented on src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java:
```diff
@@ -27,16 +28,48 @@
+    private final ILivingTargetType targetType;
+    private final LivingEntity originalTarget;
+    
     public LivingSetAttackTargetEvent(LivingEntity entity, LivingEntity target)
```
Deprecate, mark for removal.

Budschie replied:
Fixed: d8371baa7a87a75a24e43080473f41f7cb0d75ba

ChampionAsh5357 reviewed (changes requested):
What is the purpose of the `LivingSetAttackTargetEvent` now? It seems like you can just check whether there is a change through `LivingChangeTargetEvent` by scheduling on a lower priority.

Budschie committed:
Added @Cancelable annotation to LivingChangeTargetEvent

Budschie commented:
> What is the purpose of the `LivingSetAttackTargetEvent` now? It seems like you can just check whether there is a change through `LivingChangeTargetEvent` by scheduling on a lower priority.

The purpose of the `LivingSetAttackTargetEvent` is mostly to provide binary compatibility to previous Forge versions to avoid introducing breaking changes. Thus, I should probably deprecate the event as it can be considered obsolete because of the `LivingChangeTargetEvent`.

Budschie committed:
Finished JavaDoc (I think)

Budschie committed:
Deprecated LivingSetAttackTargetEvent

Budschie committed:
Moved ILivingTargetType and LivingTargetType

Moved ILivingTargetType interface and LivingTargetType enum from
LivingSetAttackTargetEvent to LivingChangeTargetEvent.

Budschie committed:
Merge branch '1.19.x' into 1.19.x-livingsetattacktargetevent-brain-compat

ChampionAsh5357 commented on src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java:
```diff
@@ -7,36 +7,64 @@
  * <br>
  * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
  **/
+@Deprecated
```
Need to provide version number of deprecation (1.19.2) and set `forRemoval` to true.

Budschie replied:
Out of interest, if I wanted to port this PR back to 1.18.2, should I just use 1.19.2 as a version of deprecation?

ChampionAsh5357 replied:
You would change the version of deprecation in the backport to 1.18.2.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -268,6 +270,19 @@ public static void onLivingSetAttackTarget(LivingEntity entity, LivingEntity tar
         MinecraftForge.EVENT_BUS.post(new LivingSetAttackTargetEvent(entity, target));
```
Need to deprecate the above method as well since the type is no longer present.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -268,6 +270,19 @@ public static void onLivingSetAttackTarget(LivingEntity entity, LivingEntity tar
         MinecraftForge.EVENT_BUS.post(new LivingSetAttackTargetEvent(entity, target));
     }
 
+    public static void onLivingSetAttackTarget(LivingEntity entity, LivingEntity target, ILivingTargetType targetType)
```
If event is marked for removal, methods should be as well.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/world/entity/Mob.java.patch:
```diff
@@ -1,10 +1,15 @@
++      net.minecraftforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.minecraftforge.common.ForgeHooks.onLivingChangeTarget(this, p_21544_, net.minecraftforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.MOB_TARGET);
++      if(!changeTargetEvent.isCanceled()) {
++          this.f_21362_ = changeTargetEvent.getNewTarget();
++          net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, this.f_21362_);
```
Since this is being marked for removal, should add a `TODO: Remove 1.20`

ChampionAsh5357 commented on patches/minecraft/net/minecraft/world/entity/ai/behavior/StartAttacking.java.patch:
```diff
@@ -0,0 +1,16 @@
++      if(!changeTargetEvent.isCanceled()) {
++          p_24214_.m_6274_().m_21879_(MemoryModuleType.f_26372_, changeTargetEvent.getNewTarget());
++          p_24214_.m_6274_().m_21936_(MemoryModuleType.f_26326_);
++          net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(p_24214_, changeTargetEvent.getNewTarget(), net.minecraftforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.BEHAVIOR_TARGET);
```
Since this is being marked for removal, should add a `TODO: Remove 1.20`

Budschie replied:
Just to reassure that I understood you correctly: Should I add that comment to the patch (or really any occurrence of firing the `LivingSetAttackTargetEvent`) *and* the event class or should I add this comment just to the event class?

Budschie replied:
I also have a little code style question. When adding that normal comment to the LivingSetAttackTargetEvent, should I put it above the JavaDoc or beneath it?

ChampionAsh5357 replied:
> Just to reassure that I understood you correctly: Should I add that comment to the patch (or really any occurrence of firing the LivingSetAttackTargetEvent) and the event class or should I add this comment just to the event class?

I would add the comment to the event class and calls to the event class since it's easier to double check that everything should be removed.

> I also have a little code style question. When adding that normal comment to the LivingSetAttackTargetEvent, should I put it above the JavaDoc or beneath it?

Same line as the deprecation annotation imo.

Budschie replied:
Ok, I've now added the TODOs and the deprecations with commit 6c510de6a5867b7c612eef839fbd7874dce9fcaf.

ChampionAsh5357 reviewed (changes requested):
I'll get around to testing it later.

Budschie committed:
Added TODOs and deprecation for LivingSetAttackTargetEvent and its users

ChampionAsh5357 commented on patches/minecraft/net/minecraft/world/entity/Mob.java.patch:
```diff
@@ -1,10 +1,16 @@
++      if(!changeTargetEvent.isCanceled()) {
++          this.f_21362_ = changeTargetEvent.getNewTarget();
++          // TODO: Remove in 1.20
++          net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, this.f_21362_);
```
Inline the message, it doesn't need to be a separate line to patch in:
`<code here> //TODO: ...`

Budschie replied:
I've now inlined the TODO comments in commit a79ed2d588c0d53ef17d43409000a20a4c2c9018.

Budschie committed:
Inlined TODO comment in patch files
